Add the Mint.HTTP.is_connection_message/2 guard

whatyouhide opened the pull request:
Closes #247.

whatyouhide commented on lib/mint/http.ex:
```diff
@@ -107,6 +107,44 @@ defmodule Mint.HTTP do
+                  elem(message, 1) == :erlang.map_get(:socket, conn) and
+                  ((elem(message, 0) in [:ssl, :tcp] and tuple_size(message) == 3) or
+                     (elem(message, 0) in [:ssl_closed, :tcp_closed] and tuple_size(message) == 2) or
+                     (elem(message, 0) in [:ssl_error, :tcp_error] and tuple_size(message) == 3))
```
This is just extra strictness, but why not :)

ericmj commented on lib/mint/http.ex:
```diff
@@ -107,6 +107,44 @@ defmodule Mint.HTTP do
+
+  """
+  # TODO: remove the check when we depend on Elixir 1.7+.
+  if Version.match?(System.version(), ">= 1.7.0") do
```
On hexdocs.pm the docs will always show so lets add a note in the doc text as well.

whatyouhide replied:
Good callout, completely forgot to mention this in the docs.

ericmj commented on lib/mint/http.ex:
```diff
@@ -107,6 +107,44 @@ defmodule Mint.HTTP do
+  defguard is_connection_message(conn, message)
+           when is_map(conn) and
+                  is_tuple(message) and
+                  is_atom(:erlang.map_get(:__struct__, conn)) and
```
The `map_get` will cause the guard to fail if the key does not exist. Do we want to guard it with `is_map_key`?

whatyouhide replied:
Ah good catch, always forget about that. 👍

ericmj commented on test/mint/http1/conn_test.exs:
```diff
@@ -16,6 +18,15 @@ defmodule Mint.HTTP1Test do
     assert HTTP1.stream(conn, :unknown_message) == :unknown
   end
 
+  test "Mint.HTTP.is_connection_message/2 guard works with HTTP1 connections", %{conn: conn} do
```
Need a version check on this test to fix the CI failure.

ericmj commented on test/mint/http2/conn_test.exs:
```diff
@@ -29,6 +31,17 @@ defmodule Mint.HTTP2Test do
     end
   end
 
+  describe "Mint.HTTP.is_mint_message/2" do
```
Let's remove this test and move the other to `test/mint/http_test.exs` since they test the same function?

whatyouhide replied:
They are but they're testing it with a different connection. We don't have any `Mint.HTTP` tests so for now I'd rather keep the two tests so that we can have thorough testing for both HTTP1 and HTTP2.

whatyouhide commented:
@ericmj was pretty tired when I did this yesterday, missed a bunch of things and didn't even check CI 😄 I fixed the compatibility checks, found out that some stuff requires OTP 21 which is required by Elixir 1.0 so I updated the checks. For older Elixir versions, I defined the macro to raise an argument error, I think that's the way to go :)
